twilight-gateway/src/stream.rs

```diff
@@ -51,43 +75,349 @@ pub enum StartRecommendedErrorType {
+///     .await?
+///     .filter_map(|shard_result| async move { shard_result.ok() })
+///     .collect::<Vec<_>>()
+///     .await;
+///
+/// let mut stream = ShardMessageStream::new(shards.iter_mut());
+///
+/// loop {
+///     let (shard, message) = match stream.next().await {
+///         Some(Ok((shard, message))) => (shard, message),
+///         Some(Err(source)) => {
+///             tracing::warn!(?source, "error receiving message");
+///
+///             if source.is_fatal() {
+///                 break;
+///             }
+///
+///             continue;
+///         },
+///         None => break,
+///     };
+///
+///     println!("received message on shard {}: {message:?}", shard.id());
+/// }
+/// # Ok(()) }
+/// ```
+pub struct ShardMessageStream<'a> {
+    /// Set of futures resolving to the next message of each shard.
+    futures: Rc<RefCell<FutureList<'a, Message>>>,
+}
+
+impl<'a> ShardMessageStream<'a> {
+    /// Create a new stream producing websocket messages from a set of shards.
+    pub fn new(shards: impl Iterator<Item = &'a mut Shard>) -> Self {
+        let mut this = Self {
+            futures: Rc::new(RefCell::new(FuturesUnordered::new())),
+        };
+
+        for shard in shards {
+            this.add_shard(shard);
+        }
+
+        this
+    }
+
+    /// Add a shard to the stream to produce a websocket message.
+    fn add_shard(&mut self, shard: &'a mut Shard) {
+        self.futures.borrow_mut().push(Box::pin(async {
+            let result = shard.next_message().await;
+
+            NextItemOutput { result, shard }
+        }));
+    }
+}
+
+impl<'a> Stream for ShardMessageStream<'a> {
+    type Item = Result<(ShardRef<'a>, Message), ReceiveMessageError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.as_mut();
+        let poll = this.futures.borrow_mut().poll_next_unpin(cx);
+
+        match poll {
+            Poll::Ready(Some(output)) => Poll::Ready(Some(output.result.map(|message| {
+                (
+                    ShardRef {
+                        list: ShardList::Messages(Rc::clone(&this.futures)),
+                        shard: Some(output.shard),
+                    },
+                    message,
+                )
+            }))),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// Guard dereferencing to the shard that produced an event or message.
+///
+/// Note that manually causing the destructor to [not be called] will cause the
+/// shard to not be re-inserted into the stream.
+///
+/// [not be called]: std::mem::forget
+pub struct ShardRef<'a> {
+    /// List of futures the shard will be re-inserted into when the reference is
+    /// dropped.
+    list: ShardList<'a>,
+    /// Mutable reference to the shard that produced an event or message.
+    shard: Option<&'a mut Shard>,
+}
```

Comment:
Does this not mean that you will need to handle a message completely before being able to get another message. Eg if you moved the shard into another task and had it running in a loop no other messages would be received?

Reply:
please see my reply with additional context https://github.com/twilight-rs/twilight/pull/1809#issuecomment-1200325787